contracts/src/cryptography/message_hash_utils.rs

```diff
@@ -0,0 +1,87 @@
+//! Signature message hash utilities for producing digests to be consumed by
+//! {ECDSA} recovery or signing.
+//!
+//! The library provides methods for generating a hash of a message that
+//! conforms to the https://eips.ethereum.org/EIPS/eip-191[ERC-191] and https://eips.ethereum.org/EIPS/eip-712[EIP 712]
+//! specifications.
+
+use alloc::string::String;
+
+use alloy_primitives::{bytes, keccak256, B256};
+use alloy_sol_types::{sol, SolType};
+
+type EthMessageCoder = sol! {
+    tuple(bytes, bytes32)
+};
+
+type TypeHashCoder = sol! {
+    tuple(string, bytes32, bytes32)
+};
+
+/// Returns the keccak256 digest of an ERC-191 signed data with version
+/// `0x45` (`personal_sign` messages).
+///
+/// The digest is calculated by prefixing a bytes32 `message_hash` with
+/// `"\x19Ethereum Signed Message:\n32"` and hashing the result. It corresponds
+/// with the hash signed when using the [eth_sign](https://eth.wiki/json-rpc/API#eth_sign) JSON-RPC method.
+///
+/// NOTE: The `message_hash` parameter is intended to be the result of hashing a
+/// raw message with keccak256, although any bytes32 value can be safely used
+/// because the final digest will be re-hashed.
+pub fn to_eth_signed_message_hash(message_hash: B256) -> B256 {
+    // "\x19Ethereum Signed Message:\n32" in bytes
+    let prefix =
+        bytes!("19457468657265756d205369676e6564204d6573736167653a0a3332");
+    let encoded =
+        EthMessageCoder::encode_packed(&(prefix.to_vec(), *message_hash));
+    keccak256(encoded)
+}
+
+/// Returns the keccak256 digest of an EIP-712 typed data (ERC-191 version
+/// `0x01`).
+///
+/// The digest is calculated from a `domain_separator` and a `struct_hash`, by
+/// prefixing them with `\x19\x01` and hashing the result. It corresponds to the
+/// hash signed by the [eth_signTypedData](https://eips.ethereum.org/EIPS/eip-712) JSON-RPC method as part of EIP-712.
+pub fn to_typed_data_hash(domain_separator: B256, struct_hash: B256) -> B256 {
+    let encoded = TypeHashCoder::encode_packed(&(
+        String::from("\x19\x01"),
+        *domain_separator,
+        *struct_hash,
+    ));
+    keccak256(encoded)
+}
+
+#[cfg(all(test, feature = "std"))]
+mod tests {
```

Comment:
Do the values from tests come from verified Solidity contracts @developeruche ?

Reply:
yeah, they are all values gotten from a solidity contract, would add some comments now

Reply:
It came from a solidity smart contract but it is not verified...

https://gist.github.com/developeruche/3de82874ea23fe624c666ad20117dea3